File: src/test/java/kitchenpos/menus/domain/tobe/domain/MenuProductsTest.java
package kitchenpos.menus.domain.tobe.domain;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.NullAndEmptySource;

import java.math.BigDecimal;
import java.util.Arrays;
import java.util.List;
import java.util.UUID;

import static org.assertj.core.api.Assertions.*;

class MenuProductsTest {

    private List<MenuProduct> menuProducts;

    @BeforeEach
    void setUp() {
        menuProducts = Arrays.asList(
                new MenuProduct(UUID.randomUUID(), BigDecimal.valueOf(10_000), 1L),
                new MenuProduct(UUID.randomUUID(), BigDecimal.valueOf(5_000), 2L)
        );
    }

    @DisplayName("메뉴 상품 목록을 생성한다")
    @Test
    void create() {
        assertThatCode(() -> new MenuProducts(menuProducts))
                .doesNotThrowAnyException();
    }

    @DisplayName("메뉴 상품 목록은 비어있지 않아야한다")
    @ParameterizedTest
    @NullAndEmptySource
    void invalidList(List<MenuProduct> menuProducts) {
        assertThatThrownBy(() -> new MenuProducts(menuProducts))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessage("메뉴 상품 목록은 빈 값이 아니어야 합니다. 입력 값 : " + menuProducts);
    }

    @DisplayName("메뉴 상품 목록의 합계 금액을 가져온다")
    @Test
    void getTotalAmount() {
        // given
        MenuProducts menuProducts = new MenuProducts(this.menuProducts);

        // when
        BigDecimal totalAmount = menuProducts.getTotalAmount();

        // then
        assertThat(totalAmount).isEqualTo(BigDecimal.valueOf(20_000));
    }
}

File: src/test/java/kitchenpos/menus/tobe/domain/MenuProductsTest.kt
package kitchenpos.menus.tobe.domain

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.types.shouldBeInstanceOf
import java.util.UUID

private val DEFAULT_PRODUCT_ID = UUID.randomUUID()

class MenuProductsTest : BehaviorSpec({
    Given("메뉴 상품들을 생성할 때") {
        When("메뉴 상품이 비어있지 않으면") {
            val menuProduct = MenuProduct(1, DEFAULT_PRODUCT_ID, 1000)

            Then("메뉴 상품들이 생성된다") {
                val menuProducts = MenuProducts(mutableListOf(menuProduct))
                menuProducts.shouldBeInstanceOf<MenuProducts>()
            }
        }

        When("메뉴 상품이 주어지지 않은 경우") {
            Then("예외가 발생한다") {
                shouldThrow<IllegalArgumentException> {
                    MenuProducts(mutableListOf())
                }
            }
        }
    }
})
